project.yaml

```diff
@@ -907,14 +907,43 @@ Layer:
               ('railway_' || (CASE WHEN railway IN ('platform') THEN railway ELSE NULL END)),
               (('aeroway_' || CASE WHEN aeroway IN ('runway', 'taxiway', 'helipad') THEN aeroway ELSE NULL END))
             ) AS feature
-          FROM planet_osm_polygon
-          WHERE highway IN ('residential', 'unclassified', 'pedestrian', 'service', 'footway', 'living_street', 'track', 'path', 'platform', 'services')
-            OR railway IN ('platform')
-            OR aeroway IN ('runway', 'taxiway', 'helipad')
+          FROM (
+          SELECT 
+              way, highway, railway, aeroway, z_order, way_area FROM planet_osm_polygon
+            WHERE highway IN ('residential', 'unclassified', 'pedestrian', 'service', 'footway', 'living_street', 'track', 'path', 'platform', 'services')
+              OR railway IN ('platform')
+              OR aeroway IN ('taxiway', 'helipad')
+          UNION ALL
+          SELECT -- runways with width either tagged or estimated from runway length
+              ST_Buffer(
+                way,
+                0.5*(St_Distance(ST_StartPoint(way), ST_EndPoint(way))/St_Distance_Sphere(ST_Transform(ST_StartPoint(way), 4326), ST_Transform(ST_EndPoint(way), 4326)))*
+                  CASE
+                    WHEN width ~ '^-?\d{1,4}(\.\d+)?$' THEN width::NUMERIC
```

Comment:
> Do you mean a web mercator database or rendering?

Both - although the only practically relevant case for the projection in the database differing from rendering would be EPSG:4326 in the database.

Currently AFAIK there is one case in the code assuming a metric coordinate system in the database: 

https://github.com/gravitystorm/openstreetmap-carto/blob/master/project.yaml#L726

Practically this is only a problem for EPSG:4326 databases - all other coordinate systems are generally nominally in meters.  Most relevant cases will be traverse mercator projections for local maps and polar projections like [here](http://maps.imagico.de/#map=3/87.030/72.387&lang=en&l=topo&r=arctic&ui=8).  These would work right now but not with this change in the version you propose.

> Well, the scale is the same everywhere - the data is in mercator meters, the projection is mercator, and it's the same mercator meters / px anywhere in an image. All measurements are in some projection

Scale factor refers to the ratio between projection units and ground units, exactly what you compute with the cos() term.  There is no generic postgis function for computing it so you have to do it by hand.  This differs for all conformal projections across the map area and is useful to know but usually is not really required per feature so computing it once per metatile would be sufficient.